File: test/native/drive_test.cpp
// Native drive test for the interactive bridge.
//
// It drives the decision-task fixture through the run, stop, and resume model with no clock,
// supplying the entry, choice, and exit decisions itself, and checks three things: that the
// engine stops at each decision and reports it as pending, that a submitted choice value is
// applied and appears in the log, and that re-submitting a consumed request identifier is
// rejected rather than crashing. The instance sets x to minus two, so the enumeration offered
// for the choice is the set from the model with x substituted.

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

#include "Controller.h"
#include "Engine.h"
#include "Monitor.h"

using namespace BPMNOS::WASM;
using json = nlohmann::ordered_json;

static std::string readFile(const std::string& path) {
  std::ifstream file(path, std::ios::binary);
  if (!file) {
    std::cerr << "cannot open " << path << "\n";
    std::exit(2);
  }
  std::stringstream buffer;
  buffer << file.rdbuf();
  return buffer.str();
}

static void check(bool condition, const std::string& message) {
  if (!condition) {
    std::cerr << "FAIL: " << message << "\n";
    std::exit(1);
  }
  std::cerr << "ok: " << message << "\n";
}

int main(int argc, char** argv) {
  std::string fixtureDir = (argc > 1) ? argv[1] : "test/fixtures";
  std::string modelXml = readFile(fixtureDir + "/DecisionTask_with_enumeration.bpmn");
  std::string instanceCsv =
    "INSTANCE_ID; NODE_ID; INITIALIZATION\n"
    "Instance_1; Process_1;\n"
    "Instance_1; Activity_1; x := -2\n";

  Engine engine;
  Monitor monitor;
  Controller controller;
  engine.attachMonitor(&monitor);
  engine.attachController(&controller);

  check(!engine.loadModel(modelXml).contains("error"), "loadModel");
  check(!engine.loadInstances(instanceCsv).contains("error"), "loadInstances");
  engine.configure(json{ {"provider", "static"} });

  json state = engine.start();
  check(!state.contains("error"), "start");
  std::cerr << "start snapshot: " << state.dump() << "\n";

  json fullLog = json::array();
  auto absorb = [&](const json& snapshot) {
    if (snapshot.contains("log")) {
      for (const auto& entry : snapshot["log"]) {
        fullLog.push_back(entry);
      }
    }
  };
  absorb(state);

  // Drive the engine while it is waiting for a decision. This timer-less fixture quiesces
  // once entry, choice, and exit have been supplied; reaching a formally terminal state would
  // additionally require advancing time past the instantiation horizon, which is a clock
  // concern and is not exercised here.
  double submittedChoice = 0;
  int entryCount = 0;
  int choiceCount = 0;
  int exitCount = 0;
  int guard = 0;
  while (!state["pending"].empty() && guard++ < 50) {
    const auto& decisionRequest = state["pending"][0];
    std::string type = decisionRequest["type"];
    json decision = { {"requestId", decisionRequest["requestId"]}, {"type", type} };
    if (type == "entry") {
      ++entryCount;
    }
    else if (type == "exit") {
      ++exitCount;
    }
    else if (type == "choice") {
      ++choiceCount;
      json choices = json::array();
      for (const auto& choice : decisionRequest["choices"]) {
        check(choice.contains("enumeration") && !choice["enumeration"].empty(),
              "the choice offers an enumeration of allowed values");
        double value = choice["enumeration"][0].get<double>();
        choices.push_back(value);
        submittedChoice = value;
      }
      decision["choices"] = choices;
    }
    json accepted = controller.submitDecision(decision);
    check(!accepted.contains("rejected"), "submitDecision accepted (" + type + ")");
    state = engine.resume();
    check(!state.contains("error"), "resume");
    absorb(state);
  }
  check(state["pending"].empty(), "the engine quiesced with no pending decision");
  check(entryCount == 1 && choiceCount == 1 && exitCount == 1,
        "entry, choice, and exit were each driven exactly once");

  bool applied = false;
  for (const auto& entry : fullLog) {
    if (entry.contains("token")) {
      const auto& token = entry["token"];
      if (token.value("nodeId", std::string()) == "Activity_1"
          && token.value("state", std::string()) == "COMPLETED"
          && token.contains("status") && token["status"].contains("choice")
          && token["status"]["choice"].get<double>() == submittedChoice) {
        applied = true;
      }
    }
  }
  check(applied, "the submitted choice was applied on Activity_1 at COMPLETED");

  json stale = controller.submitDecision(json{ {"requestId", 1}, {"type", "entry"} });
  check(stale.contains("rejected"), "re-submitting a consumed request is rejected, not a crash");

  std::cerr << "ALL PASSED\n";
  return 0;
}

File: test/wasm/drive_test.mjs
// WebAssembly drive test. It loads the compiled module in Node and drives the decision-task
// fixture through the same run, stop, and resume model as the native test, confirming that the
// engine actually executes inside WebAssembly: the entry, choice, and exit decisions are
// requested and supplied, the submitted choice is applied, and a stale submission is rejected.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import createBpmnos from '../../dist/bpmnos.mjs';

const here = dirname(fileURLToPath(import.meta.url));
const root = join(here, '..', '..');

function check(condition, message) {
  if (!condition) {
    console.error(`FAIL: ${message}`);
    process.exit(1);
  }
  console.error(`ok: ${message}`);
}

const module = await createBpmnos();

const modelXml = readFileSync(join(root, 'test', 'fixtures', 'DecisionTask_with_enumeration.bpmn'), 'utf8');
const instanceCsv =
  'INSTANCE_ID; NODE_ID; INITIALIZATION\n' +
  'Instance_1; Process_1;\n' +
  'Instance_1; Activity_1; x := -2\n';

const engine = new module.Engine();
const monitor = new module.Monitor();
const controller = new module.Controller();
engine.attachMonitor(monitor);
engine.attachController(controller);

check(!('error' in JSON.parse(engine.loadModel(modelXml))), 'loadModel');
check(!('error' in JSON.parse(engine.loadInstances(instanceCsv))), 'loadInstances');
engine.configure(JSON.stringify({ provider: 'static' }));

let state = JSON.parse(engine.start());
check(!('error' in state), 'start');

const fullLog = [];
const absorb = (s) => { if (s.log) for (const e of s.log) fullLog.push(e); };
absorb(state);

let submittedChoice = 0;
let entryCount = 0, choiceCount = 0, exitCount = 0, guard = 0;
while (state.pending.length > 0 && guard++ < 50) {
  const request = state.pending[0];
  const decision = { requestId: request.requestId, type: request.type };
  if (request.type === 'entry') entryCount++;
  else if (request.type === 'exit') exitCount++;
  else if (request.type === 'choice') {
    choiceCount++;
    decision.choices = request.choices.map((c) => {
      check(c.enumeration && c.enumeration.length > 0, 'the choice offers an enumeration');
      submittedChoice = c.enumeration[0];
      return c.enumeration[0];
    });
  }
  const accepted = JSON.parse(controller.submitDecision(JSON.stringify(decision)));
  check(!('rejected' in accepted), `submitDecision accepted (${request.type})`);
  state = JSON.parse(engine.resume());
  check(!('error' in state), 'resume');
  absorb(state);
}
check(state.pending.length === 0, 'the engine quiesced with no pending decision');
check(entryCount === 1 && choiceCount === 1 && exitCount === 1,
  'entry, choice, and exit were each driven exactly once');

const applied = fullLog.some((e) =>
  e.token && e.token.nodeId === 'Activity_1' && e.token.state === 'COMPLETED' &&
  e.token.status && e.token.status.choice === submittedChoice);
check(applied, 'the submitted choice was applied on Activity_1 at COMPLETED');

const stale = JSON.parse(controller.submitDecision(JSON.stringify({ requestId: 1, type: 'entry' })));
check('rejected' in stale, 'a consumed request is rejected, not a crash');

console.error('ALL PASSED (WebAssembly)');
